onnxruntime/core/framework/session_state.h

```diff
@@ -59,6 +60,93 @@ struct MemoryPatternGroup;
 class MemoryInfo;
 #endif
 
+struct PartialGraphExecutionManager {
+  PartialGraphExecutionManager() {
+    graph_runs_counter_ = 0;
+  }
+
+  int64_t GetNewPartialGraphRunId() {
+    std::lock_guard<OrtMutex> lock(lock_);
+    int64_t new_run_id;
+    if (available_run_ids_.size() > 0) {
+      new_run_id = available_run_ids_.back();
+      available_run_ids_.pop_back();
+    } else {
+      new_run_id = graph_runs_counter_++;
+    }
+
+    ORT_ENFORCE(to_be_executed_runs_.find(new_run_id) == to_be_executed_runs_.end());
+
+    to_be_executed_runs_.insert(new_run_id);
+    return new_run_id;
+  }
+
+  bool IsValidRunId(int64_t run_id) {
+    std::lock_guard<OrtMutex> lock(lock_);
+    auto it_in_progress_runs = graph_runs_.find(run_id);
+    auto it_to_be_executed = to_be_executed_runs_.find(run_id);
+
+    return ((it_in_progress_runs != graph_runs_.end()) ^ (it_to_be_executed != to_be_executed_runs_.end()));
+  }
+
+  bool IsToBeExecutedRunId(int64_t run_id) {
+    std::lock_guard<OrtMutex> lock(lock_);
+    auto it = to_be_executed_runs_.find(run_id);
+    return it != to_be_executed_runs_.end();
+  }
+
+  void CancelPartialGraphRun(int64_t run_id) {
+    //REVIEW(codemzs): Does it make sense to cancel a partial run that was not "prepared". May be?
+    std::lock_guard<OrtMutex> lock(lock_);
+    auto it = graph_runs_.find(run_id);
+
+    ORT_ENFORCE(it != graph_runs_.end());
+
+    available_run_ids_.push_back(run_id);
+    graph_runs_.erase(it);
+  }
+
+  void PreparePartialRun(std::shared_ptr<ExecutionFrame> frame, int64_t run_id, size_t program_counter) {
+    std::lock_guard<OrtMutex> lock(lock_);
+    auto it = graph_runs_.find(run_id);
+
+    ORT_ENFORCE(it == graph_runs_.end());
+
```

Comment:
you can fuse these two lines into one, if "it" is not used in the latter code. I usually find it more common to write vector.find(x) == vector.end()